### ISSUE TITLE
template cache -corruption - cache clear mechanism

### ISSUE THREAD

**Issue by tony-shannon** [ethercis/ethercis#180]
After work with IDCR - Top issues.v0 we found a problem which seems related to a corrupt template
Multiple attempts to reload the template and restart ecis-server failed

So we loaded a related duplicate template ; IDCR - Top3issues.v0 and that seems to work out
The new template seems ok just now, but we need to know how to resolve corrupt templates with flush mechanism, either as custom command or document the solution?

@freshehr has helped ID the problem, thanks for more info
(I cant remember the extra space issue Ian)

**Comment by chevalleyc** [comment]
Templates cannot be altered as soon as a related composition is stored; template are used as.. template to rebuild compositions from a serialized image in the DB (under the json format). Thus, updating/amending/purging templates with a same id (if I understood your description) is not an option as a migrate composition tool can be complex to implement. 

I need to finalize CR #133 since this will ensure referential integrity as templates are referenced internally by a UID and not by a template_id causing this type of issue.

**Comment by freshehr** [comment]
@Christian - I am not aware of this template being altered since upload. It was working fine yesterday morning then for no reason seemed to become corrupted. We tried to upload the same template, unchanged without fixing the issue. We do need to be able to upload amended templates with the same templateId as this is a very common situation during development, accepting that previously stored compositions will become inaccessible. This is something I will raise at the ADL workshop next week - we badly need to standardise template versioning but it is a tricky area.
In summary, I don't think the corrupted template issue that we saw yesterday was due to a template change.

**Comment by chevalleyc** [comment]
I have seen this template has been modified directly into the repository yesterday from 17:40 and ethecis restarted several times (from the log). I can't pretend 100% that a memory corruption (since templates are held in memory from startup) cannot happen, but it is extremely unlikely.

**Comment by chevalleyc** [comment]
Anyway, if you can submit a sequence of actions/events that can be used to reproduce the issue, I can try to fix it.

**Comment by freshehr** [comment]
Indeed but that was us trying to fix the problem. Dmitry had successfully posted a composition then it stopped working. I am pretty sure there was no interim upload of an updated template, especially as I was probably asleep!! I have a memory of there being some sort of custom command to purge the cache - I am imagining this?? Also the template that was uploaded after 1740 will have been identical.

**Comment by chevalleyc** [comment]
There is a command to update the cache, still experimental. No 'purge' to date

**Comment by chevalleyc** [comment]
I can't either see any template upload done using the REST API

**Comment by freshehr** [comment]
Tony did it via Webmin.

**Comment by freshehr** [comment]
I did an POST /template at 11.26am yesterday which was succesful but still got error messages when trying to create a composition or example composition from template.

Note that there was a typo in the templateID - an extra space character which was not in the file name. Is it possible that something might have got chewed up by that?

**Comment by chevalleyc** [comment]
I check locally, hang on

**Comment by chevalleyc** [comment]
@freshehr performing `http://ethercis1.3-test.ripple.foundation:8080/rest/v1/template/IDCR - Top issues.v0/example?format=FLAT` works. Could you give me more info about 'corrupted' please?
